File: server/sketchup_mcp/tools/booleans.py
from typing import Literal

from mcp.server.fastmcp import FastMCP

from .. import connection

BooleanOp = Literal["union", "subtract", "difference", "intersect", "intersection"]


def register(mcp: FastMCP) -> None:
    @mcp.tool()
    def boolean_operation(operation: BooleanOp, target_id: int, tool_id: int) -> dict:
        """Perform a solid boolean operation between two groups/components.

        - `union`: combine the two solids into one
        - `subtract` (alias `difference`): remove `tool_id` from `target_id`
        - `intersect` (alias `intersection`): keep only the overlap

        Both inputs are consumed and replaced by the result. The
        returned `id` refers to the resulting group.
        """
        return connection.send("boolean_operation", {
            "operation": operation,
            "target_id": target_id,
            "tool_id": tool_id,
        })

File: extension/su_mcp/tools/booleans.rb
module SU_MCP
  module Tools
    module Booleans
      OPERATIONS = {
        "union"        => :union,
        "subtract"     => :subtract,
        "difference"   => :subtract,    # alias
        "intersect"    => :intersect,
        "intersection" => :intersect,   # alias
      }.freeze

      def self.perform(params)
        op_key = params["operation"].to_s.downcase
        method = OPERATIONS[op_key]
        raise ArgumentError, "operation must be one of: #{OPERATIONS.keys.join(', ')}" unless method

        target = SU_MCP::Entities.find_solid!(params["target_id"], "target")
        tool   = SU_MCP::Entities.find_solid!(params["tool_id"],   "tool")

        result = target.public_send(method, tool)
        raise "Boolean #{op_key} produced no geometry (do the solids overlap?)" unless result

        { success: true, id: result.entityID, operation: op_key }
      end
    end
  end
end

SU_MCP::Dispatcher.register("boolean_operation") { |params| SU_MCP::Tools::Booleans.perform(params) }
